File: packages/blocks/src/ops.ts
/* Pure document operations — the kanban board.ts idiom: fresh arrays on
   change, untouched blocks keep object identity, no-ops return the same
   document reference. Mirrors crates/forge-blocks/src/ops.rs. */
import type { Block, BlockDocument, BlockColumn, TextBlock } from './types';
import { createBlock, isTextBlock, newId } from './types';

export type Parent = { kind: 'root' } | { kind: 'column'; columnsId: string; col: number };

export interface Located {
  block: Block;
  parent: Parent;
  index: number;
}

/* ---------------- Lookup ---------------------------------------------------- */

/** Every block in document order; `columns` containers contribute their cell
    children column-major (the container itself is not yielded). */
export function* iterBlocks(doc: BlockDocument): Generator<Located> {
  for (let i = 0; i < doc.blocks.length; i++) {
    const b = doc.blocks[i]!;
    if (b.type === 'columns') {
      for (let c = 0; c < b.columns.length; c++) {
        const col = b.columns[c]!;
        for (let j = 0; j < col.blocks.length; j++) {
          yield {
            block: col.blocks[j]!,
            parent: { kind: 'column', columnsId: b.id, col: c },
            index: j,
          };
        }
      }
    } else {
      yield { block: b, parent: { kind: 'root' }, index: i };
    }
  }
}

export function findBlock(doc: BlockDocument, id: string): Located | null {
  for (let i = 0; i < doc.blocks.length; i++) {
    const b = doc.blocks[i]!;
    if (b.id === id) return { block: b, parent: { kind: 'root' }, index: i };
    if (b.type === 'columns') {
      for (let c = 0; c < b.columns.length; c++) {
        const col = b.columns[c]!;
        for (let j = 0; j < col.blocks.length; j++) {
          if (col.blocks[j]!.id === id)
            return {
              block: col.blocks[j]!,
              parent: { kind: 'column', columnsId: b.id, col: c },
              index: j,
            };
        }
      }
    }
  }
  return null;
}

function listOf(doc: BlockDocument, parent: Parent): Block[] | null {
  if (parent.kind === 'root') return doc.blocks;
  const host = doc.blocks.find((b) => b.id === parent.columnsId);
  if (host?.type !== 'columns') return null;
  return host.columns[parent.col]?.blocks ?? null;
}

/** Rebuild the document with `parent`'s sibling list replaced. */
function withList(doc: BlockDocument, parent: Parent, next: Block[]): BlockDocument {
  if (parent.kind === 'root') return { ...doc, blocks: next };
  return {
    ...doc,
    blocks: doc.blocks.map((b) => {
      if (b.id !== parent.columnsId || b.type !== 'columns') return b;
      return {
        ...b,
        columns: b.columns.map((c, i) => (i === parent.col ? { ...c, blocks: next } : c)),
      };
    }),
  };
}

/* ---------------- Basic edits ------------------------------------------------ */

/** Shallow-merge a patch into the block with `id` (same type assumed). */
export function updateBlock(doc: BlockDocument, id: string, patch: object): BlockDocument {
  const loc = findBlock(doc, id);
  if (!loc) return doc;
  const list = listOf(doc, loc.parent)!;
  const next = list.slice();
  next[loc.index] = { ...loc.block, ...patch } as Block;
  return withList(doc, loc.parent, next);
}

/** Replace the block with `id` wholesale (type conversions keep the id). */
export function replaceBlock(doc: BlockDocument, id: string, block: Block): BlockDocument {
  // A columns block never nests inside a column.
  const loc = findBlock(doc, id);
  if (!loc || (block.type === 'columns' && loc.parent.kind === 'column')) return doc;
  const list = listOf(doc, loc.parent)!;
  const next = list.slice();
  next[loc.index] = block;
  return withList(doc, loc.parent, next);
}

/** Insert `block` after the block with `id` (same sibling list). */
export function insertAfter(doc: BlockDocument, id: string, block: Block): BlockDocument {
  const loc = findBlock(doc, id);
  if (!loc) return doc;
  const list = listOf(doc, loc.parent)!;
  const next = list.slice();
  next.splice(loc.index + 1, 0, block);
  return withList(doc, loc.parent, next);
}

/** Remove the block; the document never goes blockless and column cells
    never go empty (they refill with an empty paragraph). */
export function removeBlock(doc: BlockDocument, id: string): BlockDocument {
  const loc = findBlock(doc, id);
  if (!loc) return doc;
  const list = listOf(doc, loc.parent)!;
  let next = list.filter((b) => b.id !== id);
  if (next.length === 0) {
    if (loc.parent.kind === 'root') next = [createBlock('paragraph')];
    else next = [createBlock('paragraph')];
  }
  return withList(doc, loc.parent, next);
}

/** Swap the block with its sibling above (-1) or below (+1). */
export function moveBlock(doc: BlockDocument, id: string, delta: -1 | 1): BlockDocument {
  const loc = findBlock(doc, id);
  if (!loc) return doc;
  const list = listOf(doc, loc.parent)!;
  const target = loc.index + delta;
  if (target < 0 || target >= list.length) return doc;
  const next = list.slice();
  next[loc.index] = list[target]!;
  next[target] = list[loc.index]!;
  return withList(doc, loc.parent, next);
}

/* ---------------- Text ops --------------------------------------------------- */

/** Enter inside a text block: split `md` at the caret. Heading/quote/
    admonition tails become paragraphs; list items continue the list (todo
    unchecked). Enter on an empty list item converts it to a paragraph in
    place instead. Returns the id to focus (caret 0). */
export function splitTextBlock(
  doc: BlockDocument,
  id: string,
  offset: number,
): { doc: BlockDocument; focusId: string } {
  const loc = findBlock(doc, id);
  if (!loc || !isTextBlock(loc.block)) return { doc, focusId: id };
  const b = loc.block;

  if (b.type === 'list_item' && b.md === '') {
    return {
      doc: replaceBlock(doc, id, { id, type: 'paragraph', md: '' }),
      focusId: id,
    };
  }

  const head = b.md.slice(0, offset);
  const tail = b.md.slice(offset);
  const tailBlock: Block =
    b.type === 'list_item'
      ? {
          id: newId(),
          type: 'list_item',
          style: b.style,
          indent: b.indent,
          ...(b.checked !== undefined ? { checked: false } : {}),
          md: tail,
        }
      : { id: newId(), type: 'paragraph', md: tail };

  let next = updateBlock(doc, id, { md: head });
  next = insertAfter(next, id, tailBlock);
  return { doc: next, focusId: tailBlock.id };
}

/** Backspace at offset 0: append this paragraph's `md` to the previous text
    block. A previous divider is deleted instead (caret stays at 0). Returns
    null when there is nothing to merge with. Callers convert non-paragraph
    text blocks to paragraphs first (the shared keyboard rule). */
export function mergeWithPrevious(
  doc: BlockDocument,
  id: string,
): { doc: BlockDocument; focusId: string; caret: number } | null {
  const loc = findBlock(doc, id);
  if (!loc || loc.index === 0 || loc.block.type !== 'paragraph') return null;
  const list = listOf(doc, loc.parent)!;
  const prev = list[loc.index - 1]!;

  if (prev.type === 'divider') {
    const next = list.slice();
    next.splice(loc.index - 1, 1);
    return { doc: withList(doc, loc.parent, next), focusId: id, caret: 0 };
  }
  if (!isTextBlock(prev)) return null;

  const caret = prev.md.length;
  const next = list.slice();
  next[loc.index - 1] = { ...prev, md: prev.md + loc.block.md } as Block;
  next.splice(loc.index, 1);
  return { doc: withList(doc, loc.parent, next), focusId: prev.id, caret };
}

/** Tab/Shift+Tab on a list item: indent by delta, clamped 0..=5. */
export function setListIndent(doc: BlockDocument, id: string, delta: number): BlockDocument {
  const loc = findBlock(doc, id);
  if (loc?.block.type !== 'list_item') return doc;
  const indent = Math.max(0, Math.min(5, loc.block.indent + delta));
  if (indent === loc.block.indent) return doc;
  return updateBlock(doc, id, { indent });
}

/** Previous/next block in navigation order, skipping dividers. */
export function prevEditable(doc: BlockDocument, id: string): Located | null {
  return step(doc, id, -1);
}
export function nextEditable(doc: BlockDocument, id: string): Located | null {
  return step(doc, id, 1);
}

function step(doc: BlockDocument, id: string, dir: -1 | 1): Located | null {
  const flat = [...iterBlocks(doc)];
  const at = flat.findIndex((l) => l.block.id === id);
  if (at < 0) return null;
  for (let i = at + dir; i >= 0 && i < flat.length; i += dir) {
    if (flat[i]!.block.type !== 'divider') return flat[i]!;
  }
  return null;
}

/* ---------------- Columns ---------------------------------------------------- */

/** Wrap a root block into an n-column layout: the block becomes column 0's
    content, other columns start with an empty paragraph. */
export function wrapInColumns(doc: BlockDocument, id: string, n: 2 | 3 | 4): BlockDocument {
  const loc = findBlock(doc, id);
  if (!loc || loc.parent.kind !== 'root' || loc.block.type === 'columns') return doc;
  const ratio = 1 / n;
  const columns: BlockColumn[] = [{ ratio, blocks: [loc.block] }];
  for (let i = 1; i < n; i++) columns.push({ ratio, blocks: [createBlock('paragraph')] });
  const next = doc.blocks.slice();
  next[loc.index] = { id: newId(), type: 'columns', columns };
  return { ...doc, blocks: next };
}

/** Append a column (max 4); existing ratios shrink proportionally. */
export function addColumn(doc: BlockDocument, columnsId: string): BlockDocument {
  const host = doc.blocks.find((b) => b.id === columnsId);
  if (host?.type !== 'columns' || host.columns.length >= 4) return doc;
  const n = host.columns.length;
  const columns = host.columns.map((c) => ({ ...c, ratio: (c.ratio * n) / (n + 1) }));
  columns.push({ ratio: 1 / (n + 1), blocks: [createBlock('paragraph')] });
  return {
    ...doc,
    blocks: doc.blocks.map((b) => (b.id === columnsId ? { ...b, columns } : b)),
  };
}

/** Remove a column; its blocks splice into the neighbour. Removing the
    second-to-last column unwraps the survivor's blocks to the root. */
export function removeColumn(doc: BlockDocument, columnsId: string, col: number): BlockDocument {
  const idx = doc.blocks.findIndex((b) => b.id === columnsId);
  const host = doc.blocks[idx];
  if (host?.type !== 'columns' || col >= host.columns.length) return doc;

  if (host.columns.length <= 2) {
    const keep = host.columns[col === 0 ? 1 : 0]!;
    const next = doc.blocks.slice();
    next.splice(idx, 1, ...(keep.blocks.length ? keep.blocks : [createBlock('paragraph')]));
    return { ...doc, blocks: next };
  }

  const removed = host.columns[col]!;
  const into = col === 0 ? 0 : col - 1;
  const remaining = host.columns.filter((_, i) => i !== col);
  const extra = removed.ratio / remaining.length;
  const columns = remaining.map((c, i) => ({
    ...c,
    ratio: c.ratio + extra,
    blocks: i === into ? [...c.blocks, ...removed.blocks] : c.blocks,
  }));
  return {
    ...doc,
    blocks: doc.blocks.map((b) => (b.id === columnsId ? { ...b, columns } : b)),
  };
}

/** Set column ratios (normalized against their sum, min 10% each). */
export function setColumnRatios(
  doc: BlockDocument,
  columnsId: string,
  ratios: number[],
): BlockDocument {
  const host = doc.blocks.find((b) => b.id === columnsId);
  if (
    host?.type !== 'columns' ||
    ratios.length !== host.columns.length ||
    ratios.some((r) => !Number.isFinite(r) || r <= 0)
  )
    return doc;
  const sum = ratios.reduce((a, b) => a + b, 0);
  const columns = host.columns.map((c, i) => ({ ...c, ratio: Math.max(0.1, ratios[i]! / sum) }));
  return {
    ...doc,
    blocks: doc.blocks.map((b) => (b.id === columnsId ? { ...b, columns } : b)),
  };
}

/* ---------------- Tables ------------------------------------------------------ */

type TableBlock = Extract<Block, { type: 'table' }>;

function patchTable(
  doc: BlockDocument,
  id: string,
  fn: (t: TableBlock) => Partial<TableBlock> | null,
): BlockDocument {
  const loc = findBlock(doc, id);
  if (loc?.block.type !== 'table') return doc;
  const patch = fn(loc.block);
  return patch ? updateBlock(doc, id, patch) : doc;
}

/** Set one cell's inline-md source; row -1 addresses the header. */
export function tableSetCell(
  doc: BlockDocument,
  id: string,
  row: number,
  col: number,
  md: string,
): BlockDocument {
  return patchTable(doc, id, (t) => {
    if (row === -1) {
      if (col >= t.header.length || t.header[col] === md) return null;
      const header = t.header.slice();
      header[col] = md;
      return { header };
    }
    if (row >= t.rows.length || col >= (t.rows[row]?.length ?? 0)) return null;
    if (t.rows[row]![col] === md) return null;
    const rows = t.rows.slice();
    rows[row] = rows[row]!.slice();
    rows[row]![col] = md;
    return { rows };
  });
}

export function tableInsertRow(doc: BlockDocument, id: string, at: number): BlockDocument {
  return patchTable(doc, id, (t) => {
    const rows = t.rows.slice();
    rows.splice(Math.min(at, rows.length), 0, new Array(t.header.length).fill(''));
    return { rows };
  });
}

export function tableRemoveRow(doc: BlockDocument, id: string, at: number): BlockDocument {
  return patchTable(doc, id, (t) => {
    if (t.rows.length <= 1 || at >= t.rows.length) return null;
    return { rows: t.rows.filter((_, i) => i !== at) };
  });
}

export function tableInsertCol(doc: BlockDocument, id: string, at: number): BlockDocument {
  return patchTable(doc, id, (t) => {
    const pos = Math.min(at, t.header.length);
    const header = t.header.slice();
    header.splice(pos, 0, '');
    const rows = t.rows.map((r) => {
      const row = r.slice();
      row.splice(Math.min(pos, row.length), 0, '');
      return row;
    });
    return { header, rows };
  });
}

export function tableRemoveCol(doc: BlockDocument, id: string, at: number): BlockDocument {
  return patchTable(doc, id, (t) => {
    if (t.header.length <= 1 || at >= t.header.length) return null;
    return {
      header: t.header.filter((_, i) => i !== at),
      rows: t.rows.map((r) => r.filter((_, i) => i !== at)),
    };
  });
}

File: crates/forge-blocks/src/ops.rs
//! Pure document editing operations shared by the TUI and egui editors.
//!
//! Ops mutate the document in place and hand back the address (and caret,
//! where relevant) the editor should focus next. Undo is snapshot-based:
//! [`Document`] is `Clone`, so editors push a clone before each op batch.

use crate::address::Address;
use crate::schema::{Block, BlockKind, Column, Document, ListStyle, Tone};

/// A line-start markdown shortcut hit: replace the block's kind with `kind`
/// (which carries the text after the prefix); the caret moves back by
/// `prefix_len` bytes.
#[derive(Clone, Debug, PartialEq)]
pub struct Shortcut {
    pub kind: BlockKind,
    pub prefix_len: usize,
}

/// Outcome of [`merge_with_previous`]: where the caret landed.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct MergeResult {
    pub focus: Address,
    pub caret: usize,
}

/// Insert a fresh block of `kind` after `addr` (same sibling list). Returns
/// the new block's address.
pub fn insert_after(doc: &mut Document, addr: Address, kind: BlockKind) -> Option<Address> {
    let index = Document::index_in_siblings(addr);
    let list = doc.siblings_mut(addr)?;
    if index >= list.len() {
        return None;
    }
    list.insert(index + 1, Block::new(kind));
    Some(Document::with_index(addr, index + 1))
}

/// Remove the block at `addr`. Returns the address to focus next (previous
/// sibling, else the block now occupying the slot, else the previous
/// navigation stop). Keeps the document non-blockless and column cells
/// non-empty.
pub fn remove(doc: &mut Document, addr: Address) -> Option<Address> {
    let index = Document::index_in_siblings(addr);
    let list = doc.siblings_mut(addr)?;
    if index >= list.len() {
        return None;
    }
    list.remove(index);
    let len = list.len();
    doc.normalize();
    let focus = if index > 0 {
        Document::with_index(addr, index - 1)
    } else if len > 0 {
        Document::with_index(addr, 0)
    } else {
        // List emptied; normalize() refilled root, or the cell got a fresh
        // paragraph — focus slot 0 either way.
        Document::with_index(addr, 0)
    };
    // The focus slot is guaranteed valid post-normalize for cells and root.
    doc.block(focus).map(|_| focus)
}

/// Enter inside a text block: split its `md` at `caret` (byte offset).
/// Heading/quote/admonition tails become paragraphs; list items continue the
/// list (same style/indent, todo unchecked). Enter on an *empty* list item
/// converts it to a paragraph instead (returns the same address). Returns the
/// address to focus (caret 0).
pub fn split(doc: &mut Document, addr: Address, caret: usize) -> Option<Address> {
    let block = doc.block_mut(addr)?;
    if matches!(&block.kind, BlockKind::ListItem { md, .. } if md.is_empty()) {
        block.kind = BlockKind::Paragraph { md: String::new() };
        return Some(addr);
    }
    let tail_kind = match &mut block.kind {
        BlockKind::ListItem {
            style,
            indent,
            md,
            checked,
        } => {
            let tail = md.split_off(caret.min(md.len()));
            BlockKind::ListItem {
                style: *style,
                indent: *indent,
                checked: checked.map(|_| false),
                md: tail,
            }
        }
        BlockKind::Paragraph { md } => {
            let tail = md.split_off(caret.min(md.len()));
            BlockKind::Paragraph { md: tail }
        }
        BlockKind::Heading { md, .. }
        | BlockKind::Quote { md }
        | BlockKind::Admonition { md, .. } => {
            let tail = md.split_off(caret.min(md.len()));
            BlockKind::Paragraph { md: tail }
        }
        _ => return None,
    };
    insert_after(doc, addr, tail_kind)
}

/// Backspace at offset 0 between two text blocks: append this block's `md`
/// to the previous sibling's and remove this block. If the previous sibling
/// is a divider, the divider is deleted instead (caret stays at 0). Only
/// paragraphs merge *into* other blocks — callers convert non-paragraph text
/// blocks to paragraphs first (the shared keyboard rule). Returns `None`
/// when there is nothing to merge with (first block of its list).
pub fn merge_with_previous(doc: &mut Document, addr: Address) -> Option<MergeResult> {
    let index = Document::index_in_siblings(addr);
    if index == 0 {
        return None;
    }
    let prev_addr = Document::with_index(addr, index - 1);
    let prev_kind = &doc.block(prev_addr)?.kind;

    if matches!(prev_kind, BlockKind::Divider) {
        let list = doc.siblings_mut(addr)?;
        list.remove(index - 1);
        return Some(MergeResult {
            focus: prev_addr,
            caret: 0,
        });
    }
    if !prev_kind.is_text() {
        return None;
    }

    let md = match &doc.block(addr)?.kind {
        BlockKind::Paragraph { md } => md.clone(),
        _ => return None,
    };
    let prev_md = doc.block_mut(prev_addr)?.kind.md_mut()?;
    let caret = prev_md.len();
    prev_md.push_str(&md);
    doc.siblings_mut(addr)?.remove(index);
    Some(MergeResult {
        focus: prev_addr,
        caret,
    })
}

/// Alt+↑/↓: swap the block with its sibling in `dir` (-1 up, +1 down).
/// Returns the block's new address.
pub fn move_block(doc: &mut Document, addr: Address, dir: i32) -> Option<Address> {
    let index = Document::index_in_siblings(addr);
    let target = index.checked_add_signed(dir as isize)?;
    let list = doc.siblings_mut(addr)?;
    if index >= list.len() || target >= list.len() {
        return None;
    }
    list.swap(index, target);
    Some(Document::with_index(addr, target))
}

/// Replace the block's kind (slash-menu / "turn into" conversion). Refuses
/// to place a `Columns` kind inside a column cell (no nesting).
pub fn set_kind(doc: &mut Document, addr: Address, kind: BlockKind) -> bool {
    if addr.in_column() && matches!(kind, BlockKind::Columns { .. }) {
        return false;
    }
    match doc.block_mut(addr) {
        Some(block) => {
            block.kind = kind;
            true
        }
        None => false,
    }
}

/// Tab/Shift+Tab on a list item: adjust indent by `delta`, clamped 0..=5.
pub fn indent_list(doc: &mut Document, addr: Address, delta: i8) -> bool {
    if let Some(Block {
        kind: BlockKind::ListItem { indent, .. },
        ..
    }) = doc.block_mut(addr)
    {
        let next = (*indent as i8 + delta).clamp(0, 5) as u8;
        if next != *indent {
            *indent = next;
            return true;
        }
    }
    false
}

/* ---------------- Columns ---------------------------------------------- */

/// Wrap the root block at `addr` into an `n`-column layout: the block
/// becomes column 0's content, the other columns start with an empty
/// paragraph. Only root, non-`Columns` blocks wrap. Focus lands on the
/// wrapped block.
pub fn wrap_in_columns(doc: &mut Document, addr: Address, n: usize) -> Option<Address> {
    let Address::Root(i) = addr else { return None };
    if !(2..=4).contains(&n) || matches!(doc.blocks.get(i)?.kind, BlockKind::Columns { .. }) {
        return None;
    }
    let block = doc.blocks.remove(i);
    let ratio = 1.0 / n as f32;
    let mut columns = vec![Column {
        ratio,
        blocks: vec![block],
    }];
    for _ in 1..n {
        columns.push(Column {
            ratio,
            blocks: vec![Block::new(BlockKind::Paragraph { md: String::new() })],
        });
    }
    doc.blocks
        .insert(i, Block::new(BlockKind::Columns { columns }));
    Some(Address::Cell {
        root: i,
        col: 0,
        idx: 0,
    })
}

/// Append a column (max 4) to the `Columns` block at root index `root`.
/// Existing ratios shrink proportionally. Returns the new column's index.
pub fn add_column(doc: &mut Document, root: usize) -> Option<usize> {
    let BlockKind::Columns { columns } = &mut doc.blocks.get_mut(root)?.kind else {
        return None;
    };
    if columns.len() >= 4 {
        return None;
    }
    let n = columns.len() as f32;
    for col in columns.iter_mut() {
        col.ratio *= n / (n + 1.0);
    }
    columns.push(Column {
        ratio: 1.0 / (n + 1.0),
        blocks: vec![Block::new(BlockKind::Paragraph { md: String::new() })],
    });
    Some(columns.len() - 1)
}

/// Remove column `col`; its blocks splice into the previous column (or the
/// next, when removing column 0). Removing the second-to-last column unwraps
/// the survivor's blocks back to the root. Returns the address to focus.
pub fn remove_column(doc: &mut Document, root: usize, col: usize) -> Option<Address> {
    let BlockKind::Columns { columns } = &mut doc.blocks.get_mut(root)?.kind else {
        return None;
    };
    if col >= columns.len() {
        return None;
    }
    if columns.len() <= 2 {
        let keep = if col == 0 { 1 } else { 0 };
        let blocks = std::mem::take(&mut columns.get_mut(keep)?.blocks);
        doc.blocks.remove(root);
        for (offset, b) in blocks.into_iter().enumerate() {
            doc.blocks.insert(root + offset, b);
        }
        doc.normalize();
        return Some(Address::Root(root));
    }
    let removed = columns.remove(col);
    let into = if col == 0 { 0 } else { col - 1 };
    let target = columns.get_mut(into)?;
    let idx = target.blocks.len().saturating_sub(1);
    target.blocks.extend(removed.blocks);
    let n = columns.len() as f32;
    let extra = removed.ratio / n;
    for c in columns.iter_mut() {
        c.ratio += extra;
    }
    Some(Address::Cell {
        root,
        col: into,
        idx,
    })
}

/// Set the column ratios (normalized against their sum; each clamped to a
/// 10% minimum share).
pub fn set_column_ratios(doc: &mut Document, root: usize, ratios: &[f32]) -> bool {
    let Some(Block {
        kind: BlockKind::Columns { columns },
        ..
    }) = doc.blocks.get_mut(root)
    else {
        return false;
    };
    if ratios.len() != columns.len() || ratios.iter().any(|r| !r.is_finite() || *r <= 0.0) {
        return false;
    }
    let sum: f32 = ratios.iter().sum();
    for (c, r) in columns.iter_mut().zip(ratios) {
        c.ratio = (r / sum).max(0.1);
    }
    true
}

/* ---------------- Tables ------------------------------------------------ */

fn table_at(
    doc: &mut Document,
    addr: Address,
) -> Option<(&mut Vec<String>, &mut Vec<Vec<String>>)> {
    match &mut doc.block_mut(addr)?.kind {
        BlockKind::Table { header, rows } => Some((header, rows)),
        _ => None,
    }
}

/// Insert an empty row at `at` (clamped to the row count).
pub fn table_insert_row(doc: &mut Document, addr: Address, at: usize) -> bool {
    let Some((header, rows)) = table_at(doc, addr) else {
        return false;
    };
    let cols = header.len().max(1);
    let at = at.min(rows.len());
    rows.insert(at, vec![String::new(); cols]);
    true
}

pub fn table_remove_row(doc: &mut Document, addr: Address, at: usize) -> bool {
    let Some((_, rows)) = table_at(doc, addr) else {
        return false;
    };
    if at >= rows.len() || rows.len() <= 1 {
        return false;
    }
    rows.remove(at);
    true
}

/// Insert an empty column at `at` (clamped), in the header and every row.
pub fn table_insert_col(doc: &mut Document, addr: Address, at: usize) -> bool {
    let Some((header, rows)) = table_at(doc, addr) else {
        return false;
    };
    let at = at.min(header.len());
    header.insert(at, String::new());
    for row in rows.iter_mut() {
        row.insert(at.min(row.len()), String::new());
    }
    true
}

pub fn table_remove_col(doc: &mut Document, addr: Address, at: usize) -> bool {
    let Some((header, rows)) = table_at(doc, addr) else {
        return false;
    };
    if at >= header.len() || header.len() <= 1 {
        return false;
    }
    header.remove(at);
    for row in rows.iter_mut() {
        if at < row.len() {
            row.remove(at);
        }
    }
    true
}

/* ---------------- Line-start shortcuts ---------------------------------- */

/// Detect a markdown shortcut typed at the start of a paragraph. Editors call
/// this after each edit of a paragraph and, on a hit, replace the kind and
/// pull the caret back by `prefix_len`.
///
/// `# `..`#### `, `- `/`* `, `1. `/`1) `, `- [ ] `/`- [x] `/`[] `, `> `,
/// ```` ```lang ````, `---`, `:::info` (and the other tones).
pub fn line_start_shortcut(text: &str) -> Option<Shortcut> {
    // Todo items first: their prefix contains the bullet prefix.
    for (p, checked) in [("- [ ] ", false), ("- [x] ", true), ("[] ", false)] {
        if let Some(rest) = text.strip_prefix(p) {
            return Some(Shortcut {
                kind: BlockKind::ListItem {
                    style: ListStyle::Todo,
                    checked: Some(checked),
                    indent: 0,
                    md: rest.to_string(),
                },
                prefix_len: p.len(),
            });
        }
    }
    for (p, level) in [("#### ", 4u8), ("### ", 3), ("## ", 2), ("# ", 1)] {
        if let Some(rest) = text.strip_prefix(p) {
            return Some(Shortcut {
                kind: BlockKind::Heading {
                    level,
                    md: rest.to_string(),
                },
                prefix_len: p.len(),
            });
        }
    }
    for p in ["- ", "* "] {
        if let Some(rest) = text.strip_prefix(p) {
            return Some(Shortcut {
                kind: BlockKind::ListItem {
                    style: ListStyle::Bullet,
                    checked: None,
                    indent: 0,
                    md: rest.to_string(),
                },
                prefix_len: p.len(),
            });
        }
    }
    for p in ["1. ", "1) "] {
        if let Some(rest) = text.strip_prefix(p) {
            return Some(Shortcut {
                kind: BlockKind::ListItem {
                    style: ListStyle::Number,
                    checked: None,
                    indent: 0,
                    md: rest.to_string(),
                },
                prefix_len: p.len(),
            });
        }
    }
    if let Some(rest) = text.strip_prefix("> ") {
        return Some(Shortcut {
            kind: BlockKind::Quote {
                md: rest.to_string(),
            },
            prefix_len: 2,
        });
    }
    if let Some(lang) = text.strip_prefix("```") {
        if lang.chars().all(|c| c.is_ascii_alphanumeric()) {
            return Some(Shortcut {
                kind: BlockKind::Code {
                    lang: lang.to_string(),
                    code: String::new(),
                },
                prefix_len: text.len(),
            });
        }
    }
    if text == "---" {
        return Some(Shortcut {
            kind: BlockKind::Divider,
            prefix_len: 3,
        });
    }
    if let Some(rest) = text.strip_prefix(":::") {
        let tone = match rest {
            "info" => Some(Tone::Info),
            "success" => Some(Tone::Success),
            "warning" => Some(Tone::Warning),
            "danger" => Some(Tone::Danger),
            _ => None,
        };
        if let Some(tone) = tone {
            return Some(Shortcut {
                kind: BlockKind::Admonition {
                    tone,
                    title: String::new(),
                    md: String::new(),
                },
                prefix_len: text.len(),
            });
        }
    }
    None
}
